src/compiler/transformers/esnext.ts

```diff
@@ -16,9 +20,157 @@ namespace ts {
                 return node;
             }
             switch (node.kind) {
+                case SyntaxKind.PropertyAccessExpression:
+                case SyntaxKind.ElementAccessExpression:
+                case SyntaxKind.CallExpression:
+                    if (node.flags & NodeFlags.OptionalChain) {
+                        const updated = visitOptionalExpression(node as OptionalChain, /*captureThisArg*/ false);
+                        Debug.assertNotNode(updated, isSyntheticReference);
+                        return updated;
+                    }
+                    // falls through
                 default:
                     return visitEachChild(node, visitor, context);
             }
         }
+
+        function flattenChain(chain: OptionalChain) {
+            const links: OptionalChain[] = [chain];
+            while (!chain.questionDotToken && !isTaggedTemplateExpression(chain)) {
+                chain = cast(chain.expression, isOptionalChain);
+                links.unshift(chain);
+            }
+            return { expression: chain.expression, chain: links };
+        }
+
+        function visitNonOptionalParenthesizedExpression(node: ParenthesizedExpression, captureThisArg: boolean): Expression {
+            const expression = visitNonOptionalExpression(node.expression, captureThisArg);
+            if (isSyntheticReference(expression)) {
+                // `(a.b)` -> { expression `((_a = a).b)`, thisArg: `_a` }
+                // `(a[b])` -> { expression `((_a = a)[b])`, thisArg: `_a` }
+                return createSyntheticReferenceExpression(updateParen(node, expression.expression), expression.thisArg);
+            }
+            return updateParen(node, expression);
+        }
+
+        function visitNonOptionalPropertyAccessExpression(node: PropertyAccessExpression, captureThisArg: boolean): Expression {
+            if (isOptionalChain(node)) {
+                // If `node` is an optional chain, then it is the outermost chain of an optional expression.
+                return visitOptionalExpression(node, captureThisArg);
+            }
+
+            let expression = visitNode(node.expression, visitor, isExpression);
+            Debug.assertNotNode(expression, isSyntheticReference);
+
+            let thisArg: Expression | undefined;
+            if (captureThisArg) {
+                // `a.b` -> { expression: `(_a = a).b`, thisArg: `_a` }
+                thisArg = createTempVariable(hoistVariableDeclaration);
+                expression = createParen(createAssignment(thisArg, expression));
+            }
+
+            expression = updatePropertyAccess(node, expression, visitNode(node.name, visitor, isIdentifier));
+            return thisArg ? createSyntheticReferenceExpression(expression, thisArg) : expression;
+        }
+
+        function visitNonOptionalElementAccessExpression(node: ElementAccessExpression, captureThisArg: boolean): Expression {

```

Comment:
might not be possible because of performance, but it would be easier to read if the property access and element access functions were combined.